nixos/modules/virtualisation/xen-dom0.nix

```diff
@@ -157,354 +285,594 @@ in
-          time=0
-          timeout=30
-          # Wait for xenstored to actually come up, timing out after 30 seconds
-          while [ $time -lt $timeout ] && ! `${cfg.package}/bin/xenstore-read -s / >/dev/null 2>&1` ; do
-              time=$(($time+1))
-              sleep 1
-          done
-
-          # Exit if we timed out
-          if ! [ $time -lt $timeout ] ; then
-              echo "Could not start Xenstore Daemon"
-              exit 1
-          fi
-        ''}
-        echo "executing xen-init-dom0"
-        ${cfg.package}/lib/xen/bin/xen-init-dom0
+      # Increase the number of loopback devices from the default (8),
+      # which is way too small because every VM virtual disk requires a
+      # loopback device.
+      extraModprobeConfig = ''
+        options loop max_loop=64
       '';
-    };
 
-    systemd.sockets.xen-store = {
-      description = "XenStore Socket for userspace API";
-      wantedBy = [ "sockets.target" ];
-      socketConfig = {
-        ListenStream = [
-          "/var/run/xenstored/socket"
-          "/var/run/xenstored/socket_ro"
-        ];
-        SocketMode = "0660";
-        SocketUser = "root";
-        SocketGroup = "root";
+      # Xen Bootspec extension. This extension allows NixOS bootloaders to
+      # fetch the `xen.efi` path and access the `cfg.bootParams` option.
+      bootspec.extensions = {
+        "org.xenproject.bootspec.v1" = {
+          xen = cfg.efi.path;
+          xenParams = cfg.bootParams;
+        };
       };
-    };
 
-    systemd.services.xen-console = {
-      description = "Xen Console Daemon";
-      wantedBy = [ "multi-user.target" ];
-      after = [ "xen-store.service" ];
-      requires = [ "xen-store.service" ];
-      preStart = ''
-        mkdir -p /var/run/xen
-        ${optionalString cfg.trace "mkdir -p /var/log/xen"}
-        grep -q control_d /proc/xen/capabilities
+      # See the `xenBootBuilder` script in the main `let...in` statement of this file.
+      loader.systemd-boot.extraInstallCommands = ''
+        ${xenBootBuilderPath} ${cfg.efi.bootBuilderFlag}
       '';
-      serviceConfig = {
-        ExecStart = ''
-          ${cfg.package}/bin/xenconsoled\
-            ${optionalString ((builtins.compareVersions cfg.package.version "4.8" >= 0)) " -i"}\
-            ${optionalString cfg.trace " --log=all --log-dir=/var/log/xen"}
-        '';
-      };
     };
 
-    systemd.services.xen-qemu = {
-      description = "Xen Qemu Daemon";
-      wantedBy = [ "multi-user.target" ];
-      after = [ "xen-console.service" ];
-      requires = [ "xen-store.service" ];
-      serviceConfig.ExecStart = ''
-        ${cfg.package-qemu}/${cfg.package-qemu.qemu-system-i386} \
-           -xen-attach -xen-domid 0 -name dom0 -M xenpv \
-           -nographic -monitor /dev/null -serial /dev/null -parallel /dev/null
+    system = {
+      # Mount the /proc/xen pseudo-filesystem if we're not using the systemd stage 1 initrd.
```

Comment:
I would prefer to only support systemd stage 1 here, as we plan to make it default in 24.11 and remove the scripted one sometime after. Doesn’t make sense to spend complexity on an implementation that’s on its way out.

Reply:
Wonderful! I really didn't want that potentially broken activation script; removed.

Reply:
Oh, should we add an assertion to prevent users using the legacy script-based stage 1 from enabling Xen?

Reply:
Yeah.